src/array/cuda/rowwise_sampling_prob.cu

```diff
@@ -0,0 +1,639 @@
+__global__ void _CSRRowWiseSampleReplaceKernel(
+    const uint64_t rand_seed,
+    const int64_t num_picks,
+    const int64_t num_rows,
+    const IdType * const in_rows,
+    const IdType * const in_ptr,
+    const IdType * const in_cols,
+    const IdType * const data,
+    const FloatType * const prob,
+    const IdType * const out_ptr,
+    IdType * const out_rows,
+    IdType * const out_cols,
+    IdType * const out_idxs,
+    const IdType * const cdf_ptr, 
+    FloatType * const cdf
+) {
+  // we assign one warp per row
+  assert(blockDim.x == CTA_SIZE);
+
+  int64_t out_row = blockIdx.x * TILE_SIZE + threadIdx.y;
+  const int64_t last_row = min(static_cast<int64_t>(blockIdx.x + 1) * TILE_SIZE, num_rows);
+
+  curandStatePhilox4_32_10_t rng;
+  curand_init((rand_seed * gridDim.x + blockIdx.x) * blockDim.y + threadIdx.y, threadIdx.x, 0, &rng);
+
+  while (out_row < last_row) {
+    const int64_t row = in_rows[out_row];
+
+    const int64_t in_row_start = in_ptr[row];
+    const int64_t out_row_start = out_ptr[out_row];
+    const int64_t cdf_row_start = cdf_ptr[out_row];
+
+    const int64_t deg = in_ptr[row + 1] - in_row_start;
+    const FloatType MIN_THREAD_DATA = static_cast<FloatType>(0.0f);
+
+    if (deg > 0) {
+        // Specialize BlockScan for a 1D block of CTA_SIZE threads
+        typedef cub::BlockScan<FloatType, CTA_SIZE> BlockScan;
+        // Allocate shared memory for BlockScan
+        __shared__ typename BlockScan::TempStorage temp_storage;
+        // Initialize running total
+        BlockPrefixCallbackOp<FloatType> prefix_op(MIN_THREAD_DATA);
+
+        int64_t max = (1 + (deg - 1) / CTA_SIZE) * CTA_SIZE;
+        // Have the block iterate over segments of items
+        for (int64_t idx = threadIdx.x; idx < max; idx += CTA_SIZE)
+        {
+            // Load a segment of consecutive items that are blocked across threads
+            FloatType thread_data = idx < deg ? prob[in_row_start + idx] : MIN_THREAD_DATA;
+            if (thread_data < MIN_THREAD_DATA) {
+              thread_data = MIN_THREAD_DATA;
+            }
+            // Collectively compute the block-wide inclusive prefix sum
+            BlockScan(temp_storage).InclusiveSum(thread_data, thread_data, prefix_op);
+            __syncthreads();
+
+            // Store scanned items to cdf array
+            if (idx < deg) {
+              cdf[cdf_row_start + idx] = thread_data;
+            }
+        }
+        __syncthreads();
+        
+        for (int64_t idx = threadIdx.x; idx < num_picks; idx += CTA_SIZE) {
+            // get random value
+            FloatType sum = cdf[cdf_row_start + deg - 1]; 
+            FloatType rand = static_cast<FloatType>(curand_uniform(&rng) * sum);
+            // get the offset of the first value within cdf array which compares greater than random value. 
+            int64_t item = cub::UpperBound<FloatType* , int64_t, FloatType>(&cdf[cdf_row_start], deg, rand);
```

Comment:
I think it should be `LowerBound` here.

Reply:
I'm also wondering if we can use shared memory to make this faster. To make it, we have to use a dynamic shared memory size at `max(deg) + sizeof(temp_storage)`.

Reply:
Because InclusiveSum is used above, UpperBound is used here.
I'm not sure if shared memory will speed up, can try something.